yew/src/html/classes.rs

```diff
@@ -0,0 +1,641 @@
+        Self::from_iter(t)
+    }
+}
+
+impl<T: Into<Classes> + Clone> From<&[T]> for Classes {
+    fn from(t: &[T]) -> Self {
+        Self::from_iter(t.iter().cloned())
+    }
+}
+
+impl PartialEq for Classes {
+    fn eq(&self, other: &Self) -> bool {
+        self.set.len() == other.set.len() && self.set.iter().eq(other.set.iter())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct TestClass;
+
+    impl TestClass {
+        fn as_class(&self) -> &'static str {
+            "test-class"
+        }
+    }
+
+    impl From<TestClass> for Classes {
+        fn from(x: TestClass) -> Self {
+            Classes::from(x.as_class())
+        }
+    }
+
+    #[test]
+    fn it_is_initially_empty() {
+        let subject = Classes::new();
+        assert!(subject.is_empty());
+    }
+
+    #[test]
+    fn it_pushes_value() {
+        let mut subject = Classes::new();
+        subject.push("foo");
+        assert!(!subject.is_empty());
+        assert!(subject.contains("foo"));
+    }
+
+    #[test]
+    fn it_adds_values_via_extend() {
+        let mut other = Classes::new();
+        other.push("bar");
+        let mut subject = Classes::new();
+        subject.extend(other);
+        assert!(subject.contains("bar"));
+    }
+
+    #[test]
+    fn it_contains_both_values() {
+        let mut other = Classes::new();
+        other.push("bar");
+        let mut subject = Classes::new();
+        subject.extend(other);
+        subject.push("foo");
+        assert!(subject.contains("foo"));
+        assert!(subject.contains("bar"));
+    }
+
+    #[test]
+    fn it_splits_class_with_spaces() {
+        let mut subject = Classes::new();
+        subject.push("foo bar");
+        assert!(subject.contains("foo"));
+        assert!(subject.contains("bar"));
+    }
+
+    #[test]
+    fn push_and_contains_can_be_used_with_other_objects() {
+        let mut subject = Classes::new();
+        subject.push(TestClass);
+        let other_class: Option<TestClass> = None;
+        subject.push(other_class);
+        assert!(subject.contains(TestClass.as_class()));
+    }
+
+    #[test]
+    fn can_be_extended_with_another_class() {
+        let mut other = Classes::new();
+        other.push("foo");
+        other.push("bar");
+        let mut subject = Classes::new();
+        subject.extend(other);
+        assert!(subject.contains("foo"));
+        assert!(subject.contains("bar"));
+    }
+}
+
+#[cfg(test)]
+mod macro_tests {
+    extern crate self as yew;
+
+    use crate::html::AnyScope;
+    use crate::virtual_dom::{VDiff, VNode, VTag};
+    use crate::{classes, html, NodeRef};
+    use std::any::TypeId;
+    use std::rc::Rc;
+    #[cfg(feature = "std_web")]
+    use stdweb::web::{document, IElement, INode};
+    #[cfg(feature = "wasm_test")]
+    use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};
+
+    #[cfg(feature = "wasm_test")]
+    wasm_bindgen_test_configure!(run_in_browser);
```

Comment:
I'm not sure whether we should keep these tests like that. The entire reason for moving them over here was because `Classes` are no longer a direct part of the DOM. They behave like any other attribute now, so there's no need to test `classes!` in combination with `html!`.
I think we should simplify these tests so they no longer use `html!` or run in the browser. We can keep some integration tests over in `virtual_dom` just to make sure that the deprecated stuff still works for now.
What do you think?

Reply:
So I should:

1. add a test with `<div class=(...something...)` in virtual_dom
2. remove all the layout test (the macro is tested and there are already tests for Classes itself)

I think it sounds good

Reply:
Love how GitHub shows my comment out of context in the review below :laughing:

Reply:
Fixed in eb7b457